### PR TITLE
Adding potential impacts to remaining health indicators

### PR DESCRIPTION
**Pull request by masseyke**
The ability to add impacts to indicators was added in #84899, but impacts for all indicators other than shards availability were left empty. This commit adds potential impacts for the other indicators.

### CODE REVIEW

**Comment by masseyke** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java [hunk 1]
```diff
@@ -89,7 +91,13 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         }
-
+        List<HealthIndicatorImpact> impacts = Collections.singletonList(
+            new HealthIndicatorImpact(
+                2,
+                "Snapshots in corrupted repositories cannot be restored. Data loss is possible.",
+                List.of(ImpactArea.SEARCH)
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java:98]: Should we add a `BACKUP` impact area?  Maybe the corrupted repository should be severity 1?

I'm also wondering if we should include the repository names in the impact (similarly to how we do it in the `shards_availability` indicator) https://github.com/elastic/elasticsearch/blob/1cdb03ddb1c97b4869de33515305b7daee6cebb7/server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java#L762

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java [hunk 2]
```diff
@@ -62,11 +65,18 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         } else if (ilmMetadata.getOperationMode() != OperationMode.RUNNING) {
+            List<HealthIndicatorImpact> impacts = Collections.singletonList(
+                new HealthIndicatorImpact(
+                    3,
+                    "Indices are not being rolled over, which could lead to future instability.",
+                    List.of(ImpactArea.SEARCH)
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java:72]: Should we add a `DEPLOYMENT MANAGEMENT` impact area?

I wonder about the impact message. Rollover is an implementation detail. Maybe we shouldn't mention it in this case. 

How about?
```
Automatic index lifecycle and data retention management is disabled. The performance and stability of your system could be impacted. 
```

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorService.java [hunk 3]
```diff
@@ -62,11 +65,18 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         } else if (slmMetadata.getOperationMode() != OperationMode.RUNNING) {
+            List<HealthIndicatorImpact> impacts = Collections.singletonList(
+                new HealthIndicatorImpact(
+                    3,
+                    "Scheduled snapshots are not running, which could lead to future data loss.",
+                    List.of(ImpactArea.SEARCH)
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorService.java:72]: `BACKUP` impact area?

**Review by andreidan — COMMENTED** [review]
Thanks for adding these Keith.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java [hunk 4]
```diff
@@ -62,11 +65,18 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         } else if (ilmMetadata.getOperationMode() != OperationMode.RUNNING) {
+            List<HealthIndicatorImpact> impacts = Collections.singletonList(
+                new HealthIndicatorImpact(
+                    3,
+                    "Indices are not being rolled over, which could lead to future instability.",
```
Comment (elastichelix) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java:71]: Is there anything more specific that could be said about "future instability"? As in, could you get more specific around potential for running out of resources or data loss or something that might be coming?

Comment (masseyke) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java:71]: Just thinking through what happens when shards get too big:

- Slow query response times
- OutOfMemoryErrors
- Disks filling up (leading to maybe slow query response times at first, server crashes eventually)
- Slow recovery time on node failure (which means data might be under-replicated for longer, so slower query response times).

I'm actually not sure if all of those are still the case in 8.0. I know I had some 1 TB shards in 1.x or 2.x, and the whole cluster was just an unstable mess at that point. Maybe slow query times is the thing to highlight here? @andreidan does that list above look right? Does slow query times seem like the thing to highlight here?

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java (hunk 1 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java:98]: Do we have a working definition of impact severity levels? My thinking was that it wasn't severity 1 because the system is still responsive and usable. But it would certainly be terrifying to not have a functioning snapshot repo.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorService.java [hunk 5]
```diff
@@ -62,11 +65,18 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         } else if (slmMetadata.getOperationMode() != OperationMode.RUNNING) {
+            List<HealthIndicatorImpact> impacts = Collections.singletonList(
+                new HealthIndicatorImpact(
+                    3,
+                    "Scheduled snapshots are not running, which could lead to future data loss.",
```
Comment (Leaf-Lin) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorService.java:71]:
```suggestion
                    "Scheduled snapshots are not running.  Enable schedule snapshot management to prevent loss of data in the future.",
```
I don't think disabling SLM **leads** to future data loss 😛 , more of insurance to prevent data loss... Not sure how to best word this?

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java [hunk 6]
```diff
@@ -68,7 +71,13 @@ public void testIsYellowWhenNotRunningAndPoliciesConfigured() {
                     YELLOW,
                     "SLM is not running",
                     new SimpleHealthIndicatorDetails(Map.of("slm_status", status, "policies", 1)),
-                    Collections.emptyList(),
+                    Collections.singletonList(
+                        new HealthIndicatorImpact(
+                            3,
+                            "Scheduled snapshots are not running, which could lead to future data loss.",
```
Comment (Leaf-Lin) [inline x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java:77]: Similar to my comment above
```suggestion
                            "Scheduled snapshots are not running.  Enable schedule snapshot management to prevent loss of data in the future."
```

**Review by Leaf-Lin — COMMENTED** [review]
Just made a few suggestions around impact wording...

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java [hunk 7]
```diff
@@ -89,7 +92,17 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
                 Collections.emptyList()
             );
         }
-
+        List<HealthIndicatorImpact> impacts = Collections.singletonList(
+            new HealthIndicatorImpact(
+                1,
+                String.format(
+                    Locale.ROOT,
+                    "Snapshots in corrupted repositories %s cannot be restored. Data loss is possible.",
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java:100]: I could make the plurality dependent on how many are actually corrupted. We can have more than one repository though, so it's possible that more than one is corrupted.
My intention was to make it sound really scary. I can tone that down a little like you suggest though.

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java (hunk 6 above)
Comment (masseyke) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java:77]: Since this is the impacts block rather than user actions, how about `Scheduled snapshots are not running. There might not be backups of the data.`? Or `Scheduled snapshots are not running. There might not be backups of the data that could be used to restore if data is lost in the future.`?

File: server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java [hunk 8]
```diff
@@ -27,6 +31,7 @@ public class InstanceHasMasterHealthIndicatorService implements HealthIndicatorS
 
     private static final String INSTANCE_HAS_MASTER_GREEN_SUMMARY = "Health coordinating instance has a master node.";
     private static final String INSTANCE_HAS_MASTER_RED_SUMMARY = "Health coordinating instance does not have a master node.";
+    private static final String NO_MASTER_IMPACT = "The cluster cannot create, delete, or rebalance indices, and is likely to be unstable";
```
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java:34]: > In addition to "create", I think the cluster won't be able to accept any new writes to existing indices (append/update) either?

Ohh, you're right. Writes fail.

> All the _cat (and some other APIs) will not work.

I didn't realize this, but you're right about that, too.

> For "connected service"

Do we need to call this out separately? They only don't work because the various APIs we're already talking about don't work, right?

> Maybe somehow counterintuitively, search will continue to work when the master is missing. Is it worth pointing this out?

If you don't have a master, you probably ought to drop everything and fix that problem. There's no point in making it sound not quite so bad is there?

**Comment by masseyke** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java (hunk 4 above)
Comment (masseyke) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java]: I've updated this list based on feedback from @Leaf-Lin at https://github.com/elastic/elasticsearch/pull/86197#pullrequestreview-962768375.

**Review by andreidan — APPROVED** [review]
LGTM, thanks for working on this Keith

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java (hunk 4 above)
Comment (andreidan) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java]: ++ Keith - I think for a user the "data retention" aspect would be very important (data staying for too long in the system could breach some user data laws and such) and financially, data not moving through the tiers will have a very unpleasant impact on the bills

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java (hunk 1 above)
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java]: Ah not really for now. The severity will be very important when we'll have a global impacts section and we'll possibly only report `severity: 1` (requirements TBD)

**Review by Leaf-Lin — APPROVED** [review]
LGTM

File: x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java (hunk 6 above)
Comment (Leaf-Lin) [reply x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/slm/SlmHealthIndicatorServiceTests.java]: I know it's a bit wordy, but this feels more correct than what we had previous 👍 .

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Leaf-Lin) [reply server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java]: Could you please clarify if the check only looks at the single repository `found-snapshot` that is defined by ESS? or does it check all possible repositories from `GET _snapshot/_all`?

File: server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Leaf-Lin) [reply server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java]: >> For "connected service"

> Do we need to call this out separately? They only don't work because the various APIs we're already talking about don't work, right?

I personally would prefer to make this a bit more explicitly but I will defer this to @shubhaat .  
The question is when master is not available, all connected services like "Kibana/EnterpriseSearch/Fleet" will stop working.  Should these connected services be listed as part of the `impact area`? and should the description of impact includes a statement on the "connected services"?

File: server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java]: It checks all repositories in the cluster state, which is equivalent to `GET _snapshot/_all`. I've changed it to use "repository" if there is only one corrupted, and "repositories" if there are more than one.

File: server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/coordination/InstanceHasMasterHealthIndicatorService.java]: It makes sense to me that we'd want to tell users that. But we don't usually reference Enterprise Search in Elasticsearch code. That seems like something we'd do at the Cloud level -- it sees that Elasticsearch is broken, so it knows to tell the user that all of the applications it is hosting that are built on Elasticsearch are not working. I'll go ahead and merge this for now, and can open a separate PR about this if the consensus is that we want to start referencing applications built on top of Elasticsearch more inside of Elasticsearch.
